controllers/submission.go

```diff
@@ -50,28 +34,114 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
 		return http.StatusMethodNotAllowed, nil
 	}
 
+	var body = struct {
+		Code     string
+		Language string
+	}{}
+
+	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
+		return http.StatusBadRequest, err
+	}
+
 	p, ok := passenger.FromContext(ctx)
 	if !ok {
 		return http.StatusUnauthorized, nil
 	}
 
 	resultKey, err := datastore.DecodeKey(mux.Vars(r)["resultKey"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
 
 	if !util.HasParent(p.User, resultKey) {
 		return http.StatusBadRequest, errors.New("cannot submit answer for other users")
 	}
 
 	taskKey, err := datastore.DecodeKey(mux.Vars(r)["taskKey"])
-	// Note: When more task kinds are added, see controllers.CreateFinalResult.
-	switch taskKey.Kind() {
-	case model.CodeTaskKind:
-		return runner.HandleCodeSubmission(ctx, w, r, resultKey, taskKey)
-	// TODO(victorbalan, flowlo): Use correct kind when possible.
-	case "QuestionTask":
-		return http.StatusInternalServerError, errors.New("question submissions are not yet implemented")
-	default:
-		return http.StatusBadRequest, errors.New("Unknown submission kind.")
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+
+	var task model.Task
+	if err = datastore.Get(ctx, taskKey, &task); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	// Furthermore, the name of the GCS object is derived from the of the
+	// encapsulating Submission. To avoid race conditions, allocate an ID.
+	low, _, err := datastore.AllocateIDs(ctx, model.SubmissionKind, resultKey, 1)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	submissionKey := datastore.NewKey(ctx, model.SubmissionKind, "", low, resultKey)
```

Comment:
The object name of the GCS object containing the code is deduced from the key of the submission entity it belongs to.
And this very reference is stored inside the submission entity itself. Meaning, submission somehow contains a derivate of it's own key.
The issue is, that if I would have just allocated an incomplete key, it does not hold an identifier, and I can therefore not allocate the GCS object. So what I do here is allocate an ID, try to write the code to GCS (return on failure) and only if the code is successfully written to GCS, the submission entity is written to Datastore.

If you want to change this, you'll have to start at proposing a new naming scheme for our GCS objects.